### PR TITLE
Stage Quest artifacts before runtime fallback retries

### PR DESCRIPTION
**Pull request by KjellKod**
## Summary
Make Quest artifact preparation explicit before role execution and tighten runtime fallback behavior so write-boundary failures retry on the same runtime before cross-runtime fallback.
- Keep Quest artifacts workspace-local by default for sandboxed runs.
- Align Codex role defaults on `gpt-5.4`.

## Changes
- **Runtime helpers**:
  - Add `scripts/quest_runtime/artifacts.py` for role-to-artifact resolution, artifact preparation, and workspace-local path checks.
  - Export the new helpers from `scripts/quest_runtime/__init__.py`.
- **Claude bridge path**:
  - Pass resolved artifact paths through `scripts/quest_claude_runner.py`.
  - Update `scripts/quest_runtime/claude_runner.py` to prepare role artifacts, classify failures, and perform Tier B same-runtime retries without re-truncating artifacts.
- **Quest workflow/docs**:
  - Update `.skills/quest/delegation/workflow.md` and `.ai/quest.md` to document artifact preparation, workspace-local `.quest` paths, and the Tier A/B/C fallback ladder.
  - Update `.ai/allowlist.json` to use `models.*` and `gpt-5.4` for Codex slots.
- **Planning/tests**:
  - Add the implementation idea doc in `ideas/generic-artifact-preparation-and-runtime-fallbacks.md`.
  - Add `tests/unit/test_artifacts.py` for artifact helpers and retry behavior.

## Validation
- [x] Run `bash scripts/validate-manifest.sh` and confirm Quest files are fully covered.
- [x] Run `bash scripts/validate-quest-config.sh` and confirm config/docs validation passes.
- [x] Run `pytest -q tests/unit/test_artifacts.py tests/unit/test_quest_runtime.py` and confirm runtime helper coverage passes.
Watch for: permission/write-boundary classification should trigger same-runtime escalation before any cross-runtime fallback.

## Notes
- This draft is focused on Quest runtime behavior and workflow contracts; end-to-end verification of external shared `.quest` paths is still best validated manually.

<pre>
     ▐▛███▜▌
    ▝▜█████▛▘
      ▘▘ ▝▝
Quest/Co-Authored by Claude, GPT-5.4 in Collaboration with KjellKod
</pre>

### CODE REVIEW

**Comment by KjellKod** [comment]
I found a false-negative in the new Claude bridge probe logic.

In `scripts/quest_runtime/claude_runner.py`, `run_bridge_probe()` still computes success as `cp.returncode == 0 and handoff_state == "found"`.

That does not match the runtime semantics used by `run_claude_role()`, which already treats a found handoff as success even if the bridge process times out while exiting. In a real run from the `legion` repo, the probe wrote both `probe_artifact.txt` and valid `probe_handoff.json`, but `quest_claude_probe.py` still returned:

```json
{"exit_code":1,"handoff_state":"found","result_kind":"handoff_json","source":"handoff_json","stderr":"Timed out after 30.0s","stdout":""}
```

That makes the bridge look unavailable even though the artifact contract succeeded.

I think the bug is here:
- if `handoff_state == "found"`, the probe should normalize to success the same way the role runner does, regardless of `cp.returncode`
- otherwise the probe and the real role path disagree on what "bridge available" means

I’d also add a unit test for: bridge subprocess returns non-zero / timeout-like stderr, but the handoff file exists and is valid => probe result should still be success.

----
update: GPT-5.4 comment
----
Fixed on `codex-artifact-staging` in `ce60892`, and then tightened further in `d4429fd` while addressing the later text-fallback review. `run_bridge_probe()` now normalizes any valid handoff file to success, matching `run_claude_role()` semantics, and the runtime tests include the timeout-like / non-zero exit regression case.

File: scripts/quest_claude_runner.py [hunk 1]
```diff
@@ -7,6 +7,7 @@
 import json
 from pathlib import Path
 
+from quest_runtime.artifacts import expected_artifacts_for_role
```
[the remark this answers is not in the document]
Comment (KjellKod) [reply scripts/quest_claude_runner.py:10]: Addressed in 2b2bf5f. I added the Quest runtime package files to .quest-manifest and extended scripts/validate-manifest.sh to cover scripts/quest_runtime/*.py, so the installer/upgrade path now ships the new import dependency instead of relying on local repo state. This was a real installability bug.

File: scripts/quest_runtime/artifacts.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+def expected_artifacts_for_role(
+    quest_dir: str | Path,
+    phase: str,
+    agent: str,
+    quest_mode: str = "workflow",
+) -> list[Path]:
+    """Return absolute artifact paths for the requested role invocation."""
+
+    del phase  # Role drives artifact paths; workflow phase naming is caller-facing only.
```
[the remark this answers is not in the document]
Comment (KjellKod) [reply scripts/quest_runtime/artifacts.py]: Addressed in 2b2bf5f. I kept the helper signature and added explicit role/phase validation via ROLE_PHASE_ALIASES, plus regression coverage for invalid phase mismatches. That keeps the call sites stable while making bad routing fail fast instead of silently preparing the wrong artifact set. I agree this was worth tightening.

File: scripts/quest_runtime/claude_runner.py [hunk 3]
```diff
@@ -326,6 +379,51 @@ def run_claude_role(
         stderr=stderr,
```
[the remark this answers is not in the document]
Comment (KjellKod) [reply scripts/quest_runtime/claude_runner.py:456]: I agree this is still a live bug on the current head. Right now text_fallback returns before write-boundary or permission classification, so a bridge run can satisfy the textual fallback path and skip the Tier B same-runtime retry that the workflow contract now describes. I have not fixed this one yet. My intended change is the minimal one: classify the post-run result first, route Tier B or Tier C when needed, and only use text fallback after retry routing is exhausted.

**Comment by KjellKod** [comment]
Follow-up on issuecomment-4083419223: fixed in ce60892. run_bridge_probe() now normalizes any valid handoff file to success, matching run_claude_role() semantics. I also added a regression test for the exact false-negative case: non-zero or timeout-like subprocess exit plus valid probe_handoff.json still reports bridge availability.

File: scripts/quest_runtime/claude_runner.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (KjellKod) [reply scripts/quest_runtime/claude_runner.py:456]: ----
update: GPT-5.4 comment
----
Fixed in d4429fd. `run_claude_role()` no longer short-circuits to `text_fallback` before retry routing. Tier B/Tier C classification now happens first, and text fallback remains the last-resort path after retry routing is exhausted. I also added regression coverage in `tests/unit/test_quest_runtime.py` for the permission-style first pass plus textual handoff case.

File: scripts/quest_runtime/claude_runner.py [hunk 4]
```diff
@@ -297,7 +351,68 @@ def run_claude_role(
...
+            combined_stderr = retry_note
+            if retry_result.stderr:
+                combined_stderr = f"{retry_note}\n{retry_result.stderr}"
+            return RunResult(
+                exit_code=retry_result.exit_code,
+                handoff_state=retry_result.handoff_state,
+                result_kind=retry_result.result_kind,
+                source=retry_result.source,
+                stdout=retry_result.stdout,
+                stderr=combined_stderr,
```
[the remark this answers is not in the document]
Comment (KjellKod) [reply scripts/quest_runtime/claude_runner.py:502]: ----
update: GPT-5.4 comment
----
Fixed in 087c4fa. I kept the helper default conservative and enabled Tier C explicitly from `scripts/quest_claude_runner.py` by passing `allow_text_fallback=True`. That preserves the documented bridge compatibility path without widening the low-level helper default. I also added a regression test in `tests/unit/test_quest_runtime.py` that exercises the runner entrypoint and verifies it opts into text fallback.

File: .ai/allowlist.json [hunk 5]
```diff
@@ -18,7 +18,10 @@
     "code_review": true,
     "fix_loop": false
   },
-  "review_mode": "full",
+  "arbiter": {
+    "tool": "claude",
+    "notes": "Set to 'codex' to use gpt-5.4 as Arbiter instead of Claude. Default: claude."
```
[the remark this answers is not in the document]
Comment (KjellKod) [reply .ai/allowlist.json]: ----
update: GPT-5.4 comment
----
Fixed in 0197564. I removed the stray `arbiter.tool` block from `.ai/allowlist.json` and the matching `arbiter.tool` mention from `.ai/quest.md`, so `models.arbiter` is again the single source of truth for arbiter runtime selection. I chose the smaller fix rather than adding precedence rules on top of two competing keys.

File: scripts/quest_runtime/claude_runner.py [hunk 6]
```diff
@@ -297,7 +351,68 @@ def run_claude_role(
 
     handoff_state = classify_handoff_file(resolved_handoff_file)
     text_handoff = extract_text_handoff(stdout)
-    if text_handoff is not None:
+
```
[the remark this answers is not in the document]
Comment (KjellKod) [reply scripts/quest_runtime/claude_runner.py:428]: ----
update: GPT-5.4 comment
----
Fixed in 0197564. `run_claude_role()` now normalizes any found handoff to `result_kind="handoff_json"`, even if the bridge process only yielded that handoff after the timeout/termination path. That keeps `exit_code`, `source`, and `result_kind` consistent with the bridge probe and avoids contradictory success/failure signals. I also added a regression test in `tests/unit/test_quest_runtime.py` for the timeout-plus-valid-handoff case.

File: scripts/quest_runtime/claude_runner.py [hunk 7]
```diff
@@ -297,7 +351,70 @@ def run_claude_role(
...
+    result_kind = (
```
[the remark this answers is not in the document]
Comment (KjellKod) [reply scripts/quest_runtime/claude_runner.py:429]: ----
update: GPT-5.4 comment
----
Fixed in 1c5b2a7. `run_claude_role()` now only treats a parseable `handoff.json` as success when all prepared `artifact_paths` are also non-empty. If the handoff exists but a required artifact is still empty, the run stays in failure routing instead of silently succeeding. I also added regression coverage for both the incomplete-artifact case and the Tier A/Tier B external-artifact-dir boundary, plus late `handoff_json` context-health logging.

File: scripts/quest_claude_runner.py [hunk 8]
```diff
@@ -29,6 +30,11 @@ def parse_args() -> argparse.Namespace:
 
 def main() -> int:
     args = parse_args()
+    artifact_paths = expected_artifacts_for_role(
```
[the remark this answers is not in the document]
Comment (KjellKod) [reply scripts/quest_claude_runner.py]: ----
update: GPT-5.4 comment
----
Fixed in ebe9353. `scripts/quest_claude_runner.py` now catches `ValueError` from `expected_artifacts_for_role(...)` and returns a structured JSON payload with `result_kind="invocation_error"`, `handoff_state="missing"`, and a non-zero `exit_code` instead of crashing with a traceback. I also added regression coverage in `tests/unit/test_quest_runtime.py` for the bad phase / structured error path.

File: scripts/quest_runtime/claude_runner.py [hunk 9]
```diff
@@ -215,19 +262,27 @@ def run_claude_role(
...
     exit_grace_seconds: float = 2.0,
 ) -> RunResult:
     resolved_quest_dir = resolve_path(cwd, quest_dir)
     resolved_prompt_file = resolve_path(cwd, prompt_file)
     resolved_handoff_file = resolve_path(cwd, handoff_file)
+    resolved_artifact_paths = [resolve_path(cwd, path) for path in artifact_paths or []]
+    local_artifact_paths, _ = check_artifact_paths(resolved_artifact_paths, Path(cwd).resolve())
+    if resolved_artifact_paths and not permission_escalation:
+        prepare_artifact_files(resolved_artifact_paths)
     default_add_dirs = [
```
[the remark this answers is not in the document]
Comment (KjellKod) [reply scripts/quest_runtime/claude_runner.py:342]: ----
update: GPT-5.4 comment
----
Fixed in a1ae404. `run_claude_role()` now catches `OSError` from pre-run artifact preparation and keeps it inside the structured result/fallback contract instead of throwing a traceback. For pre-run write-boundary or permission-style failures, it reuses the Tier B retry path; otherwise it returns a structured `invocation_error` result. I also added regression coverage in `tests/unit/test_artifacts.py` for both the structured-error case and the pre-run permission retry path.
